### PR TITLE
bpo-40280: Add --with-emscripten-target to build for browser or node (GH-30552)

### PR DESCRIPTION
**Pull request by tiran**
Co-authored-by: Ethan Smith <ethan@ethanhs.me>
Signed-off-by: Christian Heimes <christian@python.org>

https://bugs.python.org/issue40280

### CODE REVIEW

File: configure [hunk 1]
```diff
@@ -6223,9 +6263,13 @@ esac
...
-  case $ac_sys_system in #(
-  Emscripten) :
-    EXEEXT=.wasm ;; #(
+  case $ac_sys_system/$ac_sys_emscripten_target in #(
+  Emscripten/browser) :
+    EXEEXT=.html ;; #(
+  Emscripten/node) :
+    EXEEXT=.js ;; #(
+  wasi/*) :
+    EXEEXT=.js ;; #(
```
Comment (emmatyping) [inline configure]: I think this should be .wasm, since WASI doesn't require JS shims for system APIs.

File: configure.ac [hunk 2]
```diff
@@ -1062,6 +1062,24 @@ AS_CASE([$ac_sys_system/$ac_sys_release],
...
 
+AC_MSG_CHECKING([for --with-emscripten-target])
+AC_ARG_WITH([emscripten-target],
+  [AS_HELP_STRING([--with-emscripten-target=@<:@browser|node@:>@], [Emscripten platform])],
+[
+	AS_VAR_IF([ac_sys_system], [Emscripten], [
+    AS_CASE([$with_emscripten_target],
+      [browser], [ac_sys_emscripten_target=browser],
+      [node], [ac_sys_emscripten_target=node],
+      [AC_MSG_ERROR([Invalid argument: --with-emscripten-target=browser|node])]
```
Comment (emmatyping) [inline configure.ac:1073]: I seem to be able to run configure without passing this at all, but we should probably check that one of these is set if we are targeting `wasm32-unknown-emscripten`

**Review by emmatyping — APPROVED** [review]
Two suggestions but otherwise this looks good! It is a bit unfortunate that running `emmake make` after the correct configure doesn't "just work" but alas.

File: configure.ac (hunk 2 above)
Comment (tiran) [reply configure.ac:1073]: The option defaults for ``browser``:
```
AS_VAR_IF([ac_sys_system], [Emscripten], [ac_sys_emscripten_target=browser])
```

**Comment by tiran** [comment]
> Two suggestions but otherwise this looks good! It is a bit unfortunate that running `emmake make` after the correct configure doesn't "just work" but alas.

It does now :)

File: configure.ac (hunk 2 above)
Comment (emmatyping) [reply configure.ac:1073]: Ah, still getting the hang of reading autoconf, thanks!

**Pull request merged**